db.go

```diff
@@ -184,87 +151,89 @@ func (db *sqlDB) ExecContext(ctx context.Context, query string, args ...interfac
 // Ping verifies if a connection to each physical database is still alive,
 // establishing a connection if necessary.
 func (db *sqlDB) Ping() error {
-	return doParallely(db.totalConnection, func(i int) error {
-		if i < len(db.primaries) {
-			return db.primaries[i].Ping()
-		}
-
-		roIndex := i - len(db.primaries)
-		return db.replicas[roIndex].Ping()
+	errPrimaries := doParallely(len(db.primaries), func(i int) error {
+		return db.primaries[i].Ping()
+	})
+	errReplicas := doParallely(len(db.replicas), func(i int) error {
+		return db.replicas[i].Ping()
 	})
+	return multierr.Combine(errPrimaries, errReplicas)
 }
 
 // PingContext verifies if a connection to each physical database is still
 // alive, establishing a connection if necessary.
 func (db *sqlDB) PingContext(ctx context.Context) error {
-	return doParallely(db.totalConnection, func(i int) error {
-		if i < len(db.primaries) {
-			return db.primaries[i].PingContext(ctx)
-		}
-		roIndex := i - len(db.primaries)
-		return db.replicas[roIndex].PingContext(ctx)
+	errPrimaries := doParallely(len(db.primaries), func(i int) error {
+		return db.primaries[i].PingContext(ctx)
+	})
+	errReplicas := doParallely(len(db.replicas), func(i int) error {
+		return db.replicas[i].PingContext(ctx)
 	})
+	return multierr.Combine(errPrimaries, errReplicas)
 }
 
 // Prepare creates a prepared statement for later queries or executions
 // on each physical database, concurrently.
-func (db *sqlDB) Prepare(query string) (Stmt, error) {
-	stmt := &stmt{
-		db:           db,
-		loadBalancer: db.stmtLoadBalancer,
-	}
+func (db *sqlDB) Prepare(query string) (_stmt Stmt, err error) {
 	roStmts := make([]*sql.Stmt, len(db.replicas))
 	primaryStmts := make([]*sql.Stmt, len(db.primaries))
-	err := doParallely(db.totalConnection, func(i int) (err error) {
-		if i < len(db.primaries) {
-			primaryStmts[i], err = db.primaries[i].Prepare(query)
-			return err
-		}
 
-		roIndex := i - len(db.primaries)
-		roStmts[roIndex], err = db.replicas[roIndex].Prepare(query)
+	errPrimaries := doParallely(len(db.primaries), func(i int) (err error) {
+		primaryStmts[i], err = db.primaries[i].Prepare(query)
+		return
+	})
+	errReplicas := doParallely(len(db.replicas), func(i int) (err error) {
+		roStmts[i], err = db.replicas[i].Prepare(query)
```

Comment:
This is not good, this will force the library to prepare the statement for every query both for Write and Read ops.

The ideal way is, when is a writing ops, only primaryDB prepares the stmt. And when it's read ops, only replicas are prepared.

Reply:
Actually, the previous one also has the same problem. I think we can continue this

Reply:
Actual we can't deduce from the statement whether its a query or exec. So we are better way preparing the statement for every db

Reply:
Yes, can ignore my comment here